File: apps/server/src/queue/worker.ts
import { taskQueue, TaskJobData, Job } from "./queue";
import { logger } from "../lib/logger";
import { db } from "../lib/firestore";
import { dispatchPipeline } from "../agents/agent.dispatcher";
import { bootstrapAgents } from "../agents/agent.service";
import { runCicdPipeline } from "../modules/cicd/cicd.service";

// register all agents before the queue starts processing
bootstrapAgents();

async function processJob(data: TaskJobData): Promise<void> {
  const { taskId, projectId, title } = data;
  const startedAt = Date.now();

  logger.info("Task job started", { taskId, projectId, title });

  const taskDoc = await db.collection("tasks").doc(taskId).get();
  if (!taskDoc.exists) {
    logger.error("Task not found in Firestore — skipping", { taskId });
    return;
  }

  const task = taskDoc.data()!;
  if (task.status === "COMPLETED") {
    logger.warn("Task already completed, skipping", { taskId });
    return;
  }

  await db.collection("tasks").doc(taskId).update({
    status: "IN_PROGRESS",
    updatedAt: new Date().toISOString(),
  });

  const pipelineResults = await dispatchPipeline(taskId);
  const allPassed = pipelineResults.every((r) => r.status === "COMPLETED");

  // auto-trigger CI/CD when the full pipeline passes
  if (allPassed) {
    runCicdPipeline(projectId, taskId).catch((err) =>
      logger.error("CI/CD auto-trigger failed", { taskId, error: (err as Error).message }),
    );
  }

  const durationMs = Date.now() - startedAt;
  logger.info("Task job finished", { taskId, status: allPassed ? "COMPLETED" : "FAILED", durationMs });
}

// listen for jobs added to the in-memory queue and process them
taskQueue.on("job:added", async (job: Job) => {
  taskQueue.updateJob(job.id, { state: "active", processedOn: Date.now() });
  logger.info("Job active", { jobId: job.id, taskId: job.data.taskId });

  try {
    await processJob(job.data);
    taskQueue.updateJob(job.id, { state: "completed", finishedOn: Date.now() });
    logger.info("Job completed", { jobId: job.id, taskId: job.data.taskId });
  } catch (err) {
    const errorMsg = (err as Error).message;
    taskQueue.updateJob(job.id, { state: "failed", failedReason: errorMsg, finishedOn: Date.now() });
    logger.error("Job failed", { jobId: job.id, taskId: job.data.taskId, error: errorMsg });

    // mark the task as failed in Firestore so the UI reflects it
    await db.collection("tasks").doc(job.data.taskId).update({
      status: "FAILED",
      updatedAt: new Date().toISOString(),
    }).catch((dbErr) =>
      logger.error("Failed to mark task as FAILED", {
        taskId: job.data.taskId,
        error: (dbErr as Error).message,
      }),
    );
  }
});

logger.info("Task worker started (in-memory queue)");

File: apps/server-py/src/queue/worker.py
import asyncio
import time

from src.queue.queue import task_queue, Job, JobData
from src.core.database import db
from src.core.logger import logger
from src.core.utils import now_iso
from src.agents.langgraph_pipeline import run_langgraph_pipeline


async def _queue_dependent_tasks(completed_task_id: str, project_id: str) -> None:
    """Queue tasks that depend on the completed task."""
    all_tasks = db.collection("tasks").where("projectId", "==", project_id).stream()
    
    for task_doc in all_tasks:
        task_data = task_doc.to_dict()
        depends_on = task_data.get("dependsOn", [])
        
        # Check if this task depends on the completed task
        if completed_task_id in depends_on:
            # Check if all dependencies are completed
            all_deps_completed = True
            for dep_id in depends_on:
                dep_doc = db.collection("tasks").document(dep_id).get()
                if not dep_doc.exists or dep_doc.to_dict().get("status") != "COMPLETED":
                    all_deps_completed = False
                    break
            
            # Queue if all dependencies are met and task is still pending
            if all_deps_completed and task_data.get("status") == "PENDING":
                await task_queue.add(
                    task_doc.id,
                    JobData(
                        taskId=task_doc.id,
                        projectId=project_id,
                        title=task_data.get("title", "")
                    )
                )
                logger.info(f"Queued dependent task: {task_doc.id}")


async def _process_job(job: Job) -> None:
    task_id = job.data.taskId
    project_id = job.data.projectId
    started_at = time.monotonic()

    logger.info(f"Task job started: task={task_id}")

    task_doc = db.collection("tasks").document(task_id).get()
    if not task_doc.exists:
        logger.error(f"Task not found - skipping: task={task_id}")
        return

    if task_doc.to_dict().get("status") == "COMPLETED":
        logger.warning(f"Task already completed, skipping: task={task_id}")
        return

    db.collection("tasks").document(task_id).update({"status": "IN_PROGRESS", "updatedAt": now_iso()})

    pipeline_results = await run_langgraph_pipeline(task_id)
    all_passed = all(r.status == "COMPLETED" for r in pipeline_results)

    if all_passed:
        await _queue_dependent_tasks(task_id, project_id)

    duration_ms = int((time.monotonic() - started_at) * 1000)
    status_label = "COMPLETED" if all_passed else "FAILED"
    logger.info(f"Task job finished: task={task_id} status={status_label} duration={duration_ms}ms")


async def _job_handler(job: Job) -> None:
    task_queue.update_job(job.id, state="active")
    logger.info(f"Job active: job={job.id} task={job.data.taskId}")
    try:
        await _process_job(job)
        task_queue.update_job(job.id, state="completed")
        logger.info(f"Job completed: job={job.id}")
    except Exception as err:
        error_msg = str(err)
        task_queue.update_job(job.id, state="failed", failed_reason=error_msg)
        logger.error(f"Job failed: job={job.id} error={error_msg}")
        try:
            db.collection("tasks").document(job.data.taskId).update(
                {"status": "FAILED", "updatedAt": now_iso()}
            )
        except Exception as db_err:
            logger.error(f"Failed to mark task as FAILED: {db_err}")


task_queue.on_job(_job_handler)
logger.info("Task worker registered (in-memory queue)")
